storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,23 +3643,408 @@ dberr_t fsp_sys_tablespace_validate()
+        if (!block)
+	  continue;
+	if (!inodes->insert_seg(block->page.frame + TRX_UNDO_SEG_HDR +
+                                TRX_UNDO_FSEG_HEADER))
+          err= DB_CORRUPTION;
+        block->page.unfix();
+      }
+    }
+  }
+
+  if (err == DB_SUCCESS)
+    err= fsp_table_inodes(inodes, mtr);
+  return err;
+}
+
+/** Free the extents, fragment page from the given inode
+@param space  system tablespace
+@param page_no index node page number
+@param offset  index node offset within page
+@return error code */
+static dberr_t fseg_inode_free(fil_space_t *space, uint32_t page_no,
+                               uint16_t offset)
+{
+  dberr_t err= DB_SUCCESS;
+  fsp_old_inode old_inodes_entry;
+  mtr_t mtr;
+  mtr.start();
+  mtr.x_lock_space(space);
+  sql_print_information("InnoDB: Freeing the unused segment page "
+                        UINT32PF " offset " UINT32PF, page_no,
+                        offset);
+  old_inodes_entry.insert(0, &mtr);
+  old_inodes_entry.insert(page_no, &mtr);
+
+  uint32_t old_free_len= space->free_len;
+  buf_block_t *iblock= mtr.get_already_latched(
+    page_id_t{0, page_no}, MTR_MEMO_PAGE_SX_FIX);
+  if (!iblock)
+  {
+    err= DB_ERROR;
+err_exit:
+    sql_print_warning("InnoDB: Failed to free the unused segment");
+    old_inodes_entry.restore(&mtr);
+    mtr.commit();
+    space->free_len= old_free_len;
+    return err;
+  }
+
+  fseg_inode_t *inode= iblock->page.frame + offset;
+  xdes_t *descr= fseg_get_first_extent(inode, space, &mtr, &err);
+  while (descr && err == DB_SUCCESS)
+  {
+    old_inodes_entry.insert_descr(descr, &mtr);
+    err= fseg_free_extent(inode, iblock, space,
+                          xdes_get_offset(descr), &mtr);
+    DBUG_EXECUTE_IF("unused_undo_free_fail_4",
+                    err= DB_CORRUPTION;);
+    if (err)
+      goto err_exit;
+    descr= fseg_get_first_extent(inode, space, &mtr, &err);
+  }
+
+  if (err)
+    goto err_exit;
+
+  for (uint32_t i= 0; i < FSEG_FRAG_ARR_N_SLOTS; i++)
+  {
+    uint32_t page_no = fseg_get_nth_frag_page_no(inode, i);
+    if (page_no != FIL_NULL)
+    {
+      err= fseg_free_page_low(inode, iblock, space, page_no, &mtr);
+      DBUG_EXECUTE_IF("unused_undo_free_fail_5",
+                      err= DB_CORRUPTION;);
+      if (err)
+        goto err_exit;
+    }
+  }
+
+  /* There is a possiblity that the inode page can move from
+  FSP_SEG_INODES_FULL to FSP_SEG_INODES_FREE and remove it
+  from FSEG_SEG_INODES_FREE. So better to store the prev
+  and next inode pages */
+  const fil_addr_t prev_addr=
+    flst_get_prev_addr(iblock->page.frame + FSEG_INODE_PAGE_NODE);
+  if (prev_addr.page != FIL_NULL)
+    old_inodes_entry.insert(prev_addr.page, &mtr);
+
+  const fil_addr_t next_addr=
+    flst_get_next_addr(iblock->page.frame + FSEG_INODE_PAGE_NODE);
+
+  if (next_addr.page != FIL_NULL)
+    old_inodes_entry.insert(next_addr.page, &mtr);
+
+  fsp_free_seg_inode(space, inode, iblock, &mtr);
+  mtr.commit();
```

Comment:
I do not quite understand the purpose of the `old_insert_entry.insert()` calls here. Are they really needed? If they are, why are we ignoring the return value?